content/upgrade-es-seamless.md

```diff
@@ -0,0 +1,74 @@
+Title: Upgrade Elasticsearch cluster software and hardware seamlessly
+Date: 2015-11-24
+Category: DevOps
+Tags: elasticsearch, how-to
+Slug: upgrade-es-seamlessly
+Author: John Young
+Avatar: john-young
+
+## Cluster Upgrades
+
+Our decision to begin using Elasticsearch came from a fairly typical use case for the popular distributed data store. We had hundreds of millions of pieces of content, and we wanted to support text search across them. As we set out to migrate our architecture from a relatively simple Postgres setup to include indexing all of our documents into Elasticsearch, we did what everyone does: we played around with Elasticsearch and made some educated guesses about the hardware we would need to support our needs. And, as so often happens, we underestimated.
+
+Our biggest problem was with disk space usage. Indexing our dataset took up far more space than we had anticpated, and we needed to increase the size of the attached AWS EBS volumes on our instances. Slightly less pressing than our disk usage was our heap usage, which made us want to increase the memory of each node so that we could allocate more heap space (reminder that it's [very important](https://www.elastic.co/guide/en/elasticsearch/guide/current/heap-sizing.html#compressed_oops) to keep your heap size below 30.5gb). All the while, we were several versions behind: running 1.3.2 at the time of 1.7.0's release. We figured the best course of action was to tackle all three of these problems at the same time, without any cluster downtime.
+
+Some caveats before we begin:
+ - This only applies to minor version upgrades, like 1.x to 1.x, or 2.x to 2.x. Major version upgrades, like 1.x to 2.x, are more complicated and require a full cluster restart.
+ - If you only have one master-eligible node, then taking it down will make your cluster very sad. If you're using the more recommended number of 3 master-eligible nodes, a new one will be elected seamlessly as you take each down individually for upgrades
+ - Our infrastructure is hosted on AWS, but the underlying principles are the same across any hardware cluster
+ - If you don't already have backups of your production data, do that before thinking about any of this
+
+
+## Out with the old, in with the new
+
+From a high level, our process looks like this:
+* Launch a new, larger EC2 instance based off our current node AMIs, with larger EBS volumes attached
+* Upgrade this instance's version of Elasticsearch
+* Create a new AMI for our upgraded instance
+* Join this new node to our production cluster
+* Re-allocate all shards off an older, non-upgraded node
+* Allow the cluster to rebalance itself
+* Once all nodes have been shipped off the old node, shut it down
+* Repeat this process of spawning new instances based off the upgraded AMI until all nodes have been upgraded
```

Comment:
Reword this. Perhaps number the steps and change this to "repeat steps X through Y until all nodes are upgraded"